File: utils/nightly-nightly-emoji-decoder/utils/nightly-emoji-decoder/tests/test_decoder.py
import unittest
from unittest.mock import patch

# Mock rationale: replace the internal emoji map with a deterministic test‑specific mapping
# so the test does not depend on the production dictionary.

from src.decoder import decode, _EMOJI_MAP

class TestEmojiDecoder(unittest.TestCase):
    def test_basic_decoding(self):
        self.assertEqual(decode("🚀🌕"), "rocket moon")
        self.assertEqual(decode("🧩🔧"), "puzzle wrench")

    def test_unknown_emoji(self):
        # The emoji "🦄" is not in the default map, should become "?"
        self.assertEqual(decode("🦄"), "?")

    def test_custom_mapping_via_mock(self):
        custom_map = {"🦄": "unicorn", "🌈": "rainbow"}
        with patch.object(_EMOJI_MAP, "__getitem__", side_effect=lambda k: custom_map.get(k, "?")):
            # The decoder will now use the mocked __getitem__ behavior
            self.assertEqual(decode("🦄🌈"), "unicorn rainbow")
            self.assertEqual(decode("🚀"), "?")  # original emoji not in custom map

if __name__ == "__main__":
    unittest.main()

File: node-utils/nightly-nightly-emoji-decoder-3/tests/test_decoder.js
// test_decoder.js
const assert = require("assert");
const { decodeEmojis } = require("../src/decoder");

// Mock rationale: deterministic mapping test
assert.strictEqual(decodeEmojis("🍎🍌🍒"), "ABC");

// Unknown emoji should become '?'
assert.strictEqual(decodeEmojis("🚀"), "?");

// Mixed known and unknown
assert.strictEqual(decodeEmojis("🍎🚀🍌"), "A?B");

console.log("All tests passed.");
